src/ast_editor.ml

```diff
@@ -0,0 +1,487 @@
+  in
+  let edit = WorkspaceEdit.make () in
+  WorkspaceEdit.replace edit
+    ~uri:(TextDocument.uri document)
+    ~range ~newText:content;
+  let (_ : bool Promise.t) = Workspace.applyEdit ~edit in
+  ()
+
+let open_pp_doc instance ~document =
+  let open Promise.Syntax in
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let pp_pp_str = fetch_pp_code ~document in
+  let* doc =
+    Workspace.openTextDocument
+      (`Uri
+        (Uri.parse ("post-ppx: " ^ TextDocument.fileName document ^ "?") ()))
+  in
+  Ast_editor_state.set_changes_tracking ast_editor_state document doc;
+  replace_document_content ~content:pp_pp_str ~document:doc;
+  let+ (_ : TextEditor.t) =
+    Window.showTextDocument ~document:(`TextDocument doc)
+      ~column:ViewColumn.Beside ()
+  in
+  0
+
+let reload_pp_doc instance ~document =
+  let open Promise.Syntax in
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  let visibleTextEditors = Window.visibleTextEditors () in
+  let origin_uri =
+    match
+      Ast_editor_state.find_original_doc_by_pp_uri
+        ~uri_string:(doc_string_uri ~document)
+        (Ast_editor_state.get_origin_to_pp_doc_map ast_editor_state)
+    with
+    | Some x -> x
+    | None -> failwith "Failed finding the original document URI."
+  in
+  let* original_document =
+    Workspace.openTextDocument (`Uri (Uri.parse origin_uri ()))
+  in
+  match
+    List.find visibleTextEditors ~f:(fun editor ->
+        document_eq (TextEditor.document editor) document)
+  with
+  | None -> Promise.resolve 1
+  | Some _ ->
+    Ast_editor_state.set_origin_changed ast_editor_state
+      ~key:(doc_string_uri ~document) ~data:false;
+    replace_document_content
+      ~content:(fetch_pp_code ~document:original_document)
+      ~document;
+    Promise.resolve 1
+
+let manage_choice instance choice ~document : int Promise.t =
+  let ast_editor_state = Extension_instance.ast_editor_state instance in
+  match Ppx_tools.Pp_path.project_root_path () with
+  | None ->
+    show_message `Error "%s" "Error : project root wasn't found";
+    Promise.resolve 1
+  | Some root -> (
+    let build_cmd =
+      let cwd = Path.of_string root in
+      fun () -> Cmd.run ~cwd (Cmd.Shell "eval $(opam env); dune build")
+    in
+    let rec build_project () =
+      let open Promise.Syntax in
+      let* res = build_cmd () in
+      if res.exitCode = 0 then
+        (if
+         Map.existsi
+           (Ast_editor_state.get_pp_doc_to_changed_origin_map ast_editor_state)
+           ~f:(fun ~key ~data:_ -> String.equal key (doc_string_uri ~document))
+        then
+          reload_pp_doc
+        else
+          open_pp_doc)
+          instance ~document
+      else
+        let* perror =
+          Window.showErrorMessage
+            ~message:"Building project failed, fix project errors and retry."
+            ~choices:[ ("Retry running `dune build`", 0); ("Abandon", 1) ]
```

Comment:
Instead of using integers, can we use polymorphic variant values for the different choices?

Reply:
395d876e4c5f1b797623894fe507e6e549a6ed9b